File: 39-combination-sum/39-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        # sort list
        candidates.sort()
        result = []
        curr_list = []
        curr_sum = target
        
        def helper(curr_sum, curr_list, candidates, result, start):
            if curr_sum < 0: 
                return
            
            elif curr_sum == 0:                
                result.append(list(curr_list))
                return
            
            for i in range(start, len(candidates)):
                curr_list.append(candidates[i])
                helper(curr_sum - candidates[i], curr_list, candidates, result, i)
                del curr_list[-1]
                
        helper(curr_sum, curr_list, candidates, result, 0)
            
        return result
                
            
            

File: 39-combination-sum/39-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> temp = new ArrayList<Integer>();
        helper(res, temp, candidates, target, 0, 0);
        return res;
    }
    public void helper(List<List<Integer>> res, List<Integer> temp, int[] candidates, int target, int curr, int start) {
        if(curr <= target) {
            if(curr == target) {
                List<Integer> toAdd = new ArrayList<Integer>();
                for(int i = 0; i < temp.size(); i++) {
                    toAdd.add(temp.get(i));
                }
                res.add(toAdd);
                return;
            }
            // iterate through choices:
            for(int i = start; i < candidates.length; i++) {
                temp.add(candidates[i]);
                curr += candidates[i];
                helper(res, temp, candidates, target, curr, i);
                temp.remove(temp.size() - 1);
                curr -= candidates[i];
            } 
        }
    }
}